pkg/applicationserver/io/packages/redis/registry.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/gogo/protobuf/proto"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	ttnredis "go.thethings.network/lorawan-stack/pkg/redis"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+)
+
+var (
+	errInvalidFieldmask   = errors.DefineInvalidArgument("invalid_fieldmask", "invalid fieldmask")
+	errInvalidIdentifiers = errors.DefineInvalidArgument("invalid_identifiers", "invalid identifiers")
+)
+
+// appendImplicitAssociationGetPaths appends implicit ttnpb.ApplicationPackageAssociation get paths to paths.
+func appendImplicitAssociationGetPaths(paths ...string) []string {
+	return append(append(make([]string, 0, 4+len(paths)),
+		"created_at",
+		"ids",
+		"package_name",
+		"updated_at",
+	), paths...)
+}
+
+func applyAssociationFieldMask(dst, src *ttnpb.ApplicationPackageAssociation, paths ...string) (*ttnpb.ApplicationPackageAssociation, error) {
+	if dst == nil {
+		dst = &ttnpb.ApplicationPackageAssociation{}
+	}
+	return dst, dst.SetFields(src, paths...)
+}
+
+// ApplicationPackagesRegistry is a Redis application packages registry.
+type ApplicationPackagesRegistry struct {
+	Redis *ttnredis.Client
+}
+
+func (r *ApplicationPackagesRegistry) devKey(devUID string) string {
+	return r.Redis.Key("uid", devUID)
+}
+
+func (r *ApplicationPackagesRegistry) fPortStr(fPort uint32) string {
+	return strconv.FormatUint(uint64(fPort), 16)
+}
+
+func (r *ApplicationPackagesRegistry) associationKey(devUID string, fPort string) string {
+	return r.Redis.Key("uid", devUID, fPort)
+}
+
+func (r *ApplicationPackagesRegistry) makeAssociationKeyFunc(devUID string) func(port string) string {
+	return func(port string) string {
+		return r.associationKey(devUID, port)
+	}
+}
+
+// Get implements applicationpackages.AssociationRegistry.
+func (r ApplicationPackagesRegistry) Get(ctx context.Context, ids ttnpb.ApplicationPackageAssociationIdentifiers, paths []string) (*ttnpb.ApplicationPackageAssociation, error) {
+	pb := &ttnpb.ApplicationPackageAssociation{}
+	defer trace.StartRegion(ctx, "get application package association by id").End()
+	if err := ttnredis.GetProto(r.Redis, r.associationKey(unique.ID(ctx, ids.EndDeviceIdentifiers), r.fPortStr(ids.FPort))).ScanProto(pb); err != nil {
+		return nil, err
+	}
+	return applyAssociationFieldMask(nil, pb, appendImplicitAssociationGetPaths(paths...)...)
+}
+
+// List implements applicationpackages.AssociationRegistry.
+func (r ApplicationPackagesRegistry) List(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, paths []string) ([]*ttnpb.ApplicationPackageAssociation, error) {
+	var pbs []*ttnpb.ApplicationPackageAssociation
+	devUID := unique.ID(ctx, ids)
+	dk := r.devKey(devUID)
+
+	defer trace.StartRegion(ctx, "list application package associations by device id").End()
+
+	err := r.Redis.Watch(func(tx *redis.Tx) (err error) {
+		var cmd *ttnredis.ProtosCmd
+		if limit, offset := ttnredis.PaginationLimitAndOffsetFromContext(ctx); limit != 0 {
```

Comment:
We usually use(or at least used) short `if`s only in cases where 1 value is assigned, which is immediately used. For example:
```go
if something := f(); something > someValue {
   // Handle something > someValue
}
```

The logic in this file is already difficult enough to understand, so please do not use short `if`s here.
Also L98. Please check other occurences.